metricflow/model/semantics/join_validator.py

```diff
@@ -18,6 +18,23 @@ class DataSourceIdentifierJoinType:
     right_identifier_type: IdentifierType
 
 
+@dataclass(frozen=True)
+class DataSourceIdentifierJoin:
+    """Link between an identifier and a DataSourceIdentifierJoinType."""
+
+    identifier_reference: IdentifierReference
+    join_type: DataSourceIdentifierJoinType
+
+
+@dataclass(frozen=True)
+class DataSourceJoin:
+    """Links between two data sources and the valid way to join them."""
+
+    left_data_source_reference: DataSourceReference
+    right_data_source_reference: DataSourceReference
+    identifier_joins: List[DataSourceIdentifierJoin]
```

Comment:
For the model page, we want to be able to show which identifiers need to join together to connect two data sources, and what type of join needs to be executed to join them. So that's why I put together this data structure.
But also, how are you imagining the logic would be simplified if that were removed?

Reply:
I don't recall, but I suspect when you switch to an arbitrary-depth recursive approach whatever I was thinking will likely be obvious (or obviously wrong) so I think we should just move on. :)